chalice/local/internal.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+"""
+This is intended only for `pytest-chalice`.
+
+https://github.com/studio3104/pytest-chalice
+"""
+
+import json
+import re
+
+from logging import getLogger
+from typing import Any, Callable, Dict, Optional  # noqa
+
+from chalice import Chalice  # noqa
+from chalice.config import Config
+from chalice.local import LocalGateway
+from chalice.local import EventType, HeaderType  # noqa
+
+logger = getLogger(__name__)
+
+# Python 3.4 or older don't have JSONDecodeError within json module
+try:
+    JSONDecodeError = json.JSONDecodeError  # type: ignore
+except AttributeError:
+    JSONDecodeError = ValueError  # type: ignore
+
+
+class InternalLocalGateway(LocalGateway):
```

Comment:
I think we can update `LocalGateway` to add support for a custom context.  That way we don't need to separate `InternalLocalGateway` class.

Reply:
When I considered the porting again, I felt that the context should not be overwritten in this way.
The reason why I added `InternalLocalGateway` class was, because `LocalGateway` doesn't support `CognitoUserPoolAuthorizer` as of the current latest version.
The imeplementation I added is just a workaround, I am thinking that we would better to consider how the context should be treated when considering to add `CognitoUserPoolAuthorizer` handling feature on `LocalGateway` for now.

So, I am going to update the implementations like following,

- Abolish `InternalLocalGateway` class
- Add a new parameter to stub override context into `TestHTTPClient`
  - As @jamesls suggested https://github.com/aws/chalice/pull/1193#discussion_r330903200

Reply:
I have considered carefully again about the stubbing custom context feature, but I feel like we would better not to add it at this moment.
The motivation why I added this was, because Chalice doesn't support Cognito's context for the local gateway for now, but I wanted to handle the response depending on Cognito's context. But if I imagine that we update `LocalGateway`, it would become a non-well-designed class, I guess.
I hope this will be considered again when `LocalGateway` actually support Cognito's authorization context, to make everything simple, I think it would better to stop adding it now.

Reply:
https://github.com/aws/chalice/pull/1193/commits/160f6f7f8fae549af2b26980fe7ba39fa2f60631